In [1]:
import pandas as pd
from pathlib import Path

# Define the path
# I'm in 'processed data', so go up one level (..) to reach 'data', 
# then into 'raw data'
raw_data_path = Path('..') / 'data' / 'raw data'

# Get CSV files from both folders
csv_files_2024 = list((raw_data_path / '2024').glob('*.csv'))
csv_files_2025 = list((raw_data_path / '2025').glob('*.csv'))

# Combine the lists
all_files = csv_files_2024 + csv_files_2025

# CHECK HOW MANY FILES WERE FOUND
print(f"Files in 2024: {len(csv_files_2024)}")
print(f"Files in 2025: {len(csv_files_2025)}")
print(f"Total files: {len(all_files)}")

if len(all_files) == 0:
    print("No files found! Current working directory:")
    print(Path.cwd())
    print(f"Path exists: {raw_data_path.exists()}")
else:
    # Combine into DataFrame - simple one-liner with encoding fix
    df = pd.concat(
        [pd.read_csv(f, encoding='windows-1252') for f in all_files], 
        ignore_index=True
    )

Files in 2024: 12
Files in 2025: 6
Total files: 18


Setting dates as an index and sorting them. 

In [2]:
# df['Date of Payment'] = pd.to_datetime(df['Date of Payment'], format='%d-%m-%Y')

newdf = df.sort_values('Date of Payment').reset_index(drop= True).copy()
newdf

Date of Payment                                  Expense Type  \
0         01/02/2024                          Other It Consultancy   
1         01/02/2024            Grant-in-aid To Arms Length Bodies   
2         01/02/2024            Grant-in-aid To Arms Length Bodies   
3         01/02/2024            Grant-in-aid To Arms Length Bodies   
4         01/02/2024            Grant-in-aid To Arms Length Bodies   
...              ...                                           ...   
3491      31/12/2024      Current Grants To Private Sector - Npish   
3492      31/12/2024  Capital Grants To Private Sector - Companies   
3493      31/12/2024     R&D Current Grants To Public Corporations   
3494      31/12/2024                        Faststream - Full Cost   
3495      31/12/2024                        Faststream - Full Cost   

                                           Expense Area  \
0     Dsit - Science, Innovation And Growth - Dsit -...   
1     Dsit - Science, Innovation And Growth - Dsit -...   
2     Dsit - Science, Innovation And Growth - Dsit -...   
3     Dsit - Science, Innovation And Growth - Dsit -...   
4     Dsit - Science, Innovation And Growth - Dsit -...   
...                                                 ...   
3491  Dsit - Digital And Technology Group - Dsit - C...   
3492  Dsit - Digital And Technology Group - Dsit - D...   
3493  Dsit - Science, Innovation And Growth - Dsit -...   
3494  Dsit - Corporate Services - Dsit - Human Resou...   
3495  Dsit - Corporate Services - Dsit - Human Resou...   

                                               Supplier  Transaction Number  \
0                                   Atkinsrealis Uk Ltd              567929   
1     Ukri - Engineering And Physical Sciences Resea...              566724   
2                       Ukri - Medical Research Council              566725   
3     Ukri - Biotechnology And Biological Science Re...              566726   
4     Ukri - Biotechnology And Biological Science Re...              566730   
...                                                 ...                 ...   
3491                      The Uk Cyber Security Council              647161   
3492                               University Of Surrey              647122   
3493                                 Npl Management Ltd              647057   
3494                                     Cabinet Office              647049   
3495                                     Cabinet Office              647065   

          Amount                                        Description  \
0      134394.66  Dsit - National Underground Asset Register-oth...   
1     90000000.0  Dsit - Epsrc Financing-grant-in-aid To Arms Le...   
2     36000000.0  Dsit - Mrc Financing-grant-in-aid To Arms Leng...   
3     10000000.0  Dsit - Research Councils Pension Scheme (Bbsrc...   
4     22000000.0  Dsit - Bbsrc Financing-grant-in-aid To Arms Le...   
...          ...                                                ...   
3491    75940.76  Dsit - Cyber Security Council-current Grants T...   
3492  1419423.64  Dsit- Onp-fonrc-future Open Network Research C...   
3493   261483.81  Dsit - Npl National Timing Centre-r&D Current ...   
3494     33568.0       Dsit - Devolved Admin-faststream - Full Cost   
3495     92831.0       Dsit - Devolved Admin-faststream - Full Cost   

     Supplier Post Code Supplier Type  
0              Kt18 5bw         Grant  
1               Sn2 1et        Vendor  
2               W1n 4al      Wga Only  
3               Sn2 1uh      Wga Only  
4               Sn2 1ff      Wga Only  
...                 ...           ...  
3491           Ec3a 4aa         Grant  
3492            Gu2 7xh        Vendor  
3493           Tw11 0lw        Vendor  
3494           Np10 8fz      Wga Only  
3495           Np10 8fz      Wga Only  

[3496 rows x 9 columns]

Cleaning and checking quality data through 4 steps 
- Remove / check duplicates data 
- Handel null values 
- Standradize Data 
- Remove unnecessary coulmns or rows  

* checking missing data

In [3]:

missing_data = newdf.isna().sum()
missing_data[missing_data > 0 ].sort_values(ascending=False)

Supplier Post Code    16
Description            3
dtype: int64

* Show all duplicate rows, with full details

In [4]:
# Show all duplicate rows, with full details
duplicates = newdf[newdf.duplicated(subset=['Transaction Number'], keep= False)]
duplicates.sort_values('Transaction Number')


Date of Payment                                    Expense Type  \
1197      08/01/2024  R & D Current Grants To Private Sector - Npish   
1195      08/01/2024  R & D Current Grants To Private Sector - Npish   
1201      08/01/2024  R & D Current Grants To Private Sector - Npish   
1198      08/01/2024  R & D Current Grants To Private Sector - Npish   
1193      08/01/2024  R & D Current Grants To Private Sector - Npish   
...              ...                                             ...   
2331      20/06/2025       R&D Current Grants To Public Corporations   
2832      25/06/2025       R&D Current Grants To Public Corporations   
2833      25/06/2025       R&D Current Grants To Public Corporations   
3358      30/06/2025                Cl - Cash Cfers Paid Over To Hmt   
3360      30/06/2025                Cl - Cash Cfers Paid Over To Hmt   

                                           Expense Area  \
1197  Dsit - Science, Innovation And Growth - Dsit -...   
1195  Dsit - Science, Innovation And Growth - Dsit -...   
1201  Dsit - Science, Innovation And Growth - Dsit -...   
1198  Dsit - Science, Innovation And Growth - Dsit -...   
1193  Dsit - Science, Innovation And Growth - Dsit -...   
...                                                 ...   
2331  Dsit - Science, Innovation And Growth - Dsit -...   
2832  Dsit - Science, Innovation And Growth - Dsit -...   
2833  Dsit - Science, Innovation And Growth - Dsit -...   
3358  Dsit - Digital And Technology Group - Dsit - D...   
3360  Dsit - Digital And Technology Group - Dsit - D...   

                              Supplier  Transaction Number       Amount  \
1197               The British Academy              561995    1925839.0   
1195               The British Academy              561995     127881.0   
1201  Royal Academy Of Engineering Rae              561997   3190723.32   
1198  Royal Academy Of Engineering Rae              561997     68221.22   
1193       Academy Of Medical Sciences              561998     540498.0   
...                                ...                 ...          ...   
2331                        Met Office              696974    115960.86   
2832                        Met Office              698016    120012.55   
2833                        Met Office              698016     39432.25   
3358    Consolidated Fund Account 6622              699919  48732174.44   
3360    Consolidated Fund Account 6622              699919    620966.24   

                                            Description Supplier Post Code  \
1197  Dsit - British Academy-r & D Current Grants To...           Sw1y 5ah   
1195  Dsit - British Academy Transitional Measures A...           Sw1y 5ah   
1201  Dsit - Royal Academy Of Engineering-r & D Curr...           Sw1y 5dg   
1198  Dsit - Royal Academy Of Engineering Transition...           Sw1y 5dg   
1193  Dsit - Academy Of Medical Sciences-r & D Curre...            W1b 1qh   
...                                                 ...                ...   
2331  Dsit - Met Office - Strategic Priorities Fund-...            Ex1 3pb   
2832  Dsit - Met Office - Clean Air: Analysis & Solu...            Ex1 3pb   
2833  Dsit - Met Office - Clean Air: Analysis & Solu...            Ex1 3pb   
3358  Dsit- Ofcom Cfer Receipts - Central-cl - Cash ...           Rh10 1uh   
3360  Dsit- Bduk Cfer Receipts - Central (Ofcom Subs...           Rh10 1uh   

     Supplier Type  
1197        Vendor  
1195        Vendor  
1201        Vendor  
1198        Vendor  
1193         Grant  
...            ...  
2331      Wga Only  
2832      Wga Only  
2833      Wga Only  
3358      Wga Only  
3360      Wga Only  

[417 rows x 9 columns]

* Get Transaction Numbers that appear more than once
* How many each Transaction occures 

In [5]:

duplicate_counts = newdf['Transaction Number'].value_counts()
duplicate_counts = duplicate_counts[duplicate_counts > 1]
duplicate_counts

Transaction Number
618334    7
593836    6
611418    6
593837    6
672067    6
         ..
632135    2
575233    2
575235    2
575223    2
661677    2
Name: count, Length: 173, dtype: int64

* Number of duplicates

In [6]:
duplicates = newdf[newdf.duplicated(subset=['Transaction Number'], keep=False)]
len(duplicates)

417

Validate date ranges and spot any anomalous formats.

In [7]:
# Convert to datetime and catch format errors
newdf['Date of Payment'] = pd.to_datetime(newdf['Date of Payment'], errors='coerce')

# Find invalid formats (became NaT after conversion)
print(f"Invalid date formats: {newdf['Date of Payment'].isna().sum()}")

# Check date range
print(f"Date range: {newdf['Date of Payment'].min()} to {newdf['Date of Payment'].max()}")

# Find dates outside expected range
anomalies = newdf[
    (newdf['Date of Payment'] < '2024-01-01') | 
    (newdf['Date of Payment'] > '2025-12-31')
]
print(f"Dates outside 2024-2025: {len(anomalies)}")

# Show any anomalies found
if len(anomalies) > 0:
    print(anomalies[['Date of Payment', 'Supplier', 'Amount']])

Invalid date formats: 1856
Date range: 2024-01-02 00:00:00 to 2025-12-06 00:00:00
Dates outside 2024-2025: 0


Removing null values in (dates of payment) column 

In [8]:
newdf  = newdf[newdf['Date of Payment'].notna()].copy()

* Display invalid dates

In [9]:
invalid_dates = newdf[newdf['Date of Payment'].isna()]
print(f"Invalid dates found: {len(invalid_dates)}")

Invalid dates found: 0


In [10]:
# After removing
print(f"Rows after: {len(newdf)}")
print(f"Missing dates now: {newdf['Date of Payment'].isna().sum()}")

Rows after: 1640
Missing dates now: 0


* Duplicated Transaction Numbers after removing duplicates 

In [11]:
duplicates = newdf[newdf.duplicated(subset=['Transaction Number'], keep=False)]
print(f'Duplicated Transaction Numbers after : {len(duplicates)}')

Duplicated Transaction Numbers after : 160


In [12]:
duplicate_counts = newdf['Transaction Number'].value_counts()
duplicate_counts = duplicate_counts[duplicate_counts > 1]
duplicate_counts

Transaction Number
593836    6
672067    6
694164    5
575225    4
672550    3
         ..
575233    2
575235    2
575655    2
586609    2
587725    2
Name: count, Length: 72, dtype: int64

This code takes messy money values like "£134,394.66" and cleans them up into pure numbers like 134394.66.

In [13]:

newdf['Amount'] = newdf['Amount'].astype(str).str.replace('£', '').str.replace(',', '').astype(float)
# newdf['Amount'] = pd.to_numeric(newdf['Amount'], errors='coerce')
newdf['Amount'].head()

0      134394.66
1    90000000.00
2    36000000.00
3    10000000.00
4    22000000.00
Name: Amount, dtype: float64

* Getting a copy of the cleaned datasest

In [14]:
newdf.to_csv('master_spend_cleand_data.csv', index=False, encoding='utf-8')
